quartz/components/Head.tsx

```diff
@@ -1,30 +1,172 @@
 import { JSResourceToScriptElement } from "../util/resources"
 import { QuartzComponentConstructor, QuartzComponentProps } from "./types"
+import satori, { SatoriOptions } from "satori"
+import * as fs from "fs"
+import { ImageOptions, SocialImageOptions, getSatoriFont } from "../util/imageHelper"
+import sharp from "sharp"
+import { defaultImage } from "../util/socialImage"
+import { unescapeHTML } from "../util/escape"
+
+/**
+ * Generates social image (OG/twitter standard) and saves it as `.webp` inside the public folder
+ * @param opts options for generating image
+ */
+async function generateSocialImage(opts: ImageOptions, userOpts: SocialImageOptions) {
+  const { cfg, description, fileName, fontsPromise, title } = opts
+  const fonts = await fontsPromise
+
+  // JSX that will be used to generate satori svg
+  const imageElement = userOpts.imageStructure(cfg, userOpts, title, description, fonts)
+
+  const svg = await satori(imageElement, {
+    width: userOpts.width,
+    height: userOpts.height,
+    fonts: fonts,
+  })
+
+  // Convert svg directly to webp (with additional compression)
+  const compressed = await sharp(Buffer.from(svg)).webp({ quality: 40 }).toBuffer()
+
+  // Write to file system
+  fs.writeFileSync(`${imageDir}/${fileName}.${extension}`, compressed)
+}
+
+const extension = "webp"
+const imageDir = "public/static/social-images"
+
+const defaultOptions: SocialImageOptions = {
+  colorScheme: "lightMode",
+  width: 1200,
+  height: 676,
+  imageStructure: defaultImage,
+}
 
 export default (() => {
+  let fontsPromise: Promise<SatoriOptions["fonts"]>
+
+  let fullOptions: SocialImageOptions
   function Head({ cfg, fileData, externalResources }: QuartzComponentProps) {
+    // Initialize options if not set
+    if (!fullOptions) {
+      if (typeof cfg.generateSocialImages !== "boolean") {
+        fullOptions = { ...defaultOptions, ...cfg.generateSocialImages }
+      } else {
+        fullOptions = defaultOptions
+      }
+    }
+
+    // Memoize google fonts
+    if (!fontsPromise) {
+      fontsPromise = getSatoriFont(cfg.theme.typography.header, cfg.theme.typography.body)
+    }
+
+    const slug = fileData.filePath
+    // since "/" is not a valid character in file names, replace with "-"
+    const fileName = slug?.replaceAll("/", "-")
     const title = fileData.frontmatter?.title ?? "Untitled"
-    const description = fileData.description?.trim() ?? "No description provided"
+
+    // Get file description (priority: frontmatter > fileData > default)
+    const fdDescription = fileData.description?.trim()
+    let description = ""
+    if (fdDescription) {
+      description = unescapeHTML(fdDescription)
+    }
+
+    if (fileData.frontmatter?.socialDescription) {
+      description = fileData.frontmatter?.socialDescription as string
+    } else if (fileData.frontmatter?.description) {
+      description = fileData.frontmatter?.description
+    }
+
+    if (cfg.generateSocialImages) {
```

Comment:
move this guard up a lot further, we shouldn't even try to load the satori fonts if this option isn't enabled